ecs-cli/modules/clients/aws/route53/servicediscovery_client.go

```diff
@@ -0,0 +1,138 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package servicediscovery contains functions for working with the route53 APIs
+// that back ECS Service Discovery
+package route53
+
+import (
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+)
+
+// FindPrivateNamespace returns the ID(s) of the private namespace with the given name and vpc
+func FindPrivateNamespace(name, vpc string, config *config.CommandConfig) (*string, error) {
+	region := aws.StringValue(config.Session.Config.Region)
+	var nameMatch []*string
+	err := listNamespaces(false, config, func(namespace *servicediscovery.NamespaceSummary) bool {
+		if aws.StringValue(namespace.Name) == name {
+			nameMatch = append(nameMatch, namespace.Id)
+		}
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, namespaceID := range nameMatch {
+		hasVPC, err := checkNamespaceVPC(namespaceID, vpc, region, config)
+		if err != nil {
+			return nil, err
+		}
+		if hasVPC {
+			return namespaceID, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// FindPublicNamespace returns the ID of the public namespace with the given name
+func FindPublicNamespace(name string, config *config.CommandConfig) (*string, error) {
+	var namespace *string
+	err := listNamespaces(true, config, func(n *servicediscovery.NamespaceSummary) bool {
+		if aws.StringValue(n.Name) == name {
+			namespace = n.Id
+			return false // we found it, stop the list call
+		}
+		return true
+	})
+
+	return namespace, err
+}
+
+func checkNamespaceVPC(namespaceID *string, vpc string, region string, config *config.CommandConfig) (bool, error) {
+	namespaceInfo, err := getNamespace(namespaceID, config)
+	if err != nil {
+		return false, err
+	}
+	hostedZoneID := namespaceInfo.Namespace.Properties.DnsProperties.HostedZoneId
+	hostedZone, err := getHostedZone(hostedZoneID, config)
+	if err != nil {
+		return false, err
+	}
+	for _, hostedZoneVPC := range hostedZone.VPCs {
+		// The VPC must be in the region that we will be launching the ECS Service
+		if (aws.StringValue(hostedZoneVPC.VPCId) == vpc) && (aws.StringValue(hostedZoneVPC.VPCRegion) == region) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// Private ServiceDiscovery Client that can be mocked in unit tests
+type serviceDiscoveryClient interface {
+	ListNamespacesPages(input *servicediscovery.ListNamespacesInput, fn func(*servicediscovery.ListNamespacesOutput, bool) bool) error
+	GetNamespace(input *servicediscovery.GetNamespaceInput) (*servicediscovery.GetNamespaceOutput, error)
+}
+
+var sdClient *servicediscovery.ServiceDiscovery
+
+// mockable getter function
+var getSDClient = func(config *config.CommandConfig) serviceDiscoveryClient {
+	if sdClient == nil {
+		sdClient = servicediscovery.New(config.Session)
```

Comment:
Naive question: would it ever be the case that the `sdClient` and the `route53Client` would use different session creds?

Reply:
I can't think of any such case. We also don't currently have any mechanism in the CLI to support using two pairs of creds in the same command invocation.